src/core/strategies/token.js

```diff
@@ -0,0 +1,285 @@
+            if (token.type === 'SUB' || token.type === 'NEG') {
+                // the operator is -, simply replace it by +
+                result = replaceByPositiveSign(token, index, tokens);
+            } else if (token.type === 'ADD' || token.type === 'POS') {
+                // the operator is +, simply replace it by -
+                result = replaceByNegativeSign(token);
+            } else if (nextToken) {
+                // the operator is not + or -, simply negate the value
+                result = insertNegativeSign(nextToken);
+            }
+        } else if (nextToken && (tokensHelper.isFunction(type) || token.type === 'LPAR')) {
+            // a function or a left parenthesis precedes the operand, simply negate the operand
+            result = insertNegativeSign(nextToken);
+        }
+    }
+
+    return result;
+}
+
+/**
+ * List of known strategies to apply on an expression in order to process sign change.
+ * Each strategy will return either `null` if cannot apply, or the descriptor of the change to apply.
+ * @type {tokenStrategy[]}
+ */
+export const signStrategies = [
+    /**
+     * Strategy that applies on numeric operands only
+     * @param {number} index - The index of the current token
+     * @param {token[]} tokens - The list of tokens that represent the expression
+     * @returns {tokenChange|null} - The result of the strategy: `null` if cannot apply, or the descriptor of the change
+     */
+    function strategyNumeric(index, tokens) {
+        let token = tokens[index] || null;
+        let type = tokensHelper.getType(token);
+        let result = null;
+
+        if (tokensHelper.isDigit(type) && index >= 0) {
+            // find the first token on the left of the operand
+            while (index && tokensHelper.isDigit(type)) {
+                index--;
+                token = tokens[index] || null;
+                type = tokensHelper.getType(token);
+            }
+
+            if (tokensHelper.isDigit(type) && index === 0) {
+                // the operand is the first of the expression, so the sign is implicit +, simply negate the value
+                result = insertNegativeSign(token);
+            } else {
+                // the operand is preceded by something else
+                result = applySignChange(index, tokens);
+            }
+        }
+
+        return result;
+    },
+
+    /**
+     * Strategy that applies on operators only
+     * @param {number} index - The index of the current token
+     * @param {token[]} tokens - The list of tokens that represent the expression
+     * @returns {tokenChange|null} - The result of the strategy: `null` if cannot apply, or the descriptor of the change
+     */
+    function strategyOperator(index, tokens) {
+        let token = tokens[index] || null;
+        let type = tokensHelper.getType(token);
+        let result = null;
+
+        if (tokensHelper.isOperator(type) && index >= 0) {
```

Comment:
i would swear i have seen this exactly piece of logic up above :). maybe refactor could be considered?

Reply:
yes indeed, a refactoring could apply. Actually, most of the code already exists in https://github.com/oat-sa/tao-core-ui-fe/blob/master/src/maths/calculator/plugins/modifiers/sign.js and was ported here. I rewrote a part of it, to properly include it there, but the logic remains as it. In fact, I did not want to break the feature trying to refactor the algorithms when not necessary ;)